keps/sig-cli/3805-ssa-default/README.md

```diff
@@ -0,0 +1,520 @@
+will break some people's expectations and might break some scripts too,
+but the feature is required none-the-less for its benefits, for users:
+- Users are missing out on Server-Side Apply feature because they don't
+  know they can use it
+- Having both server-side and client-side feature is confusing for users
+  who've tried the feature, and changing from one to the other can cause
+  odd behaviors
+- Some flags in kubectl are meant for this historical CSA while some
+  flags are meant for SSA, causing even more confusion
+- strategic-merge-patch (SMP) is not maintained and broken, a
+  frustrating situation for users
+
+For maintainers:
+- Many workflows need to be considered for both paradigms, making
+  maintenance of kubectl more difficult
+- kubectl has a lot of code to maintain SMP and other client-side apply
+  mechanisms. Removing the feature can greatly reduce the complexity of
+  kubectl
+- Removes some ugly server-side code meant to deal with transition from
+  client-side apply to server-side apply.
+
+### Goals
+
+The goal is to increase usage of Server-Side Apply so that users can
+benefits from the feature, mostly by turning the feature on by default.
+
+### Non-Goals
+
+Good question.
+
+## Proposal
+
+The feature consists of adding a new `auto` value to the `--server-side`
+flag for `kubectl apply` (and corresponding `kubectl diff`) and make
+this the default value. All the other values for the flag would continue
+to work as expected (for reference, `false` continues to client-side
+apply, and `true` continues to server-side apply).
+
+The meaning of `auto` goes as follows:
+- Resources continue to be fetched (GET) before-hand
+- If the resource has a kubectl `last-applied` annotation, we infer that
+  the resource is client-side applied, and we continue to client-side
+  apply that resource
+- If the resource is new (GET returns 404), the resource is server-side applied
+- If the resource already exists but doesn't have the `last-applied`
+  annotation, the resource is server-side applied
+
+### User Stories (Optional)
+
+#### Story 1
+
+User 1 starts using `kubectl` for the first time or on a new
+project, everything will always be server-side applied, they will get
+conflicts and all the good things about server-side apply.
+
+#### Story 2
+
+User 2 only uses server-side apply all the time, they may have
+`--server-side` inserted in some scripts or configuration, or they must
+remember to always include the flag. Thanks the feature, forgetting the
+flag is now doing the right thing by server-side applying all their
+resources as expected, causing less confusion and risks of messing up.
+
+#### Story 3
+
+User 3 has never used server-side apply and always used the default
+value of `--server-side`. Applying their existing resources will
+continue to work as expected. New resources, on the other hand, will at
+first apply the same but may eventually show-up with a conflict. They
```

Comment:
This introduces a potential risk, where half of the resources is SSA, and the other CSA. Imagine a scenario: you have a directory with fileA, fileB already present on the cluster, you create fileC and run apply. What happens to fileC, is it CSA-ed or SSA-ed? Based on your description it'll be SSA-ed, which IMO, would be wrong, since that directory was already CSA-ed, and I'd expect to keep it that way. I'd suggest making the decision to CSA or SSA exactly the same for the entire directory, such that in the described example it'll continue to use CSA.

Reply:
Additionally, I'd consider putting some kind of info for the user such that it's obvious to him that we're switching to SSA, or in general which apply is being used. I want to ensure this is not confusing for users.